In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

^C


### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

Traceback (most recent call last):
  File "/content/sum.py", line 7, in <module>
    a, b = map(int, input().split())
KeyboardInterrupt
^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
7
10 48 59 62 20 36 78

Overwriting input4.txt


In [ ]:
!./max_pairwise_product_long_full < input4.txt

4836


In [ ]:
!./max_pairwise_product_long_full < input4.txt > output4.txt

In [ ]:
%%writefile input5.txt
10
100 1 3 15 80 14 23 105 78 47

Writing input5.txt


In [ ]:
!./max_pairwise_product_long_full < input5.txt

10500


In [ ]:
!./max_pairwise_product_long_full < input5.txt > output5.txt

In [ ]:
%%writefile input6.txt
5
1025 4856 4540 4563 7582

Writing input6.txt


In [ ]:
!./max_pairwise_product_long_full < input6.txt

36818192


In [ ]:
!./max_pairwise_product_long_full < input6.txt > output6.txt

In [ ]:
%%writefile input7.txt
15
856 1 785 65 96 35 11 45 458 956 702 18 63 58 77

Writing input7.txt


In [ ]:
!./max_pairwise_product_long_full < input7.txt

818336


In [ ]:
!./max_pairwise_product_long_full < input7.txt > output7.txt

In [ ]:
%%writefile input8.txt
20
0 896 362 452 12 36 543 786 5555 444 22 1 778 8885 2 4254 21 606 60 3

Writing input8.txt


In [ ]:
!./max_pairwise_product_long_full < input8.txt

49356175


In [ ]:
!./max_pairwise_product_long_full < input8.txt > output8.txt

In [ ]:
%%writefile input9.txt
2
4785962 8964521

Writing input9.txt


In [ ]:
!./max_pairwise_product_long_full < input9.txt

1428534458


In [ ]:
!./max_pairwise_product_long_full < input9.txt > output9.txt

In [ ]:
%%writefile input10.txt
8
78572 410275 1 75757 3663 4254543 4125 653

Writing input10.txt


In [ ]:
!./max_pairwise_product_long_full < input10.txt

1775907149


In [ ]:
!./max_pairwise_product_long_full < input10.txt > output10.txt

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

^C


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

IndentationError: ignored

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

589



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n= 10

/bin/bash: line 1: input4.txt: No such file or directory



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


^C


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

306/200001000010

1.5299923499617505e-09

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.861s
user	0m0.850s
sys	0m0.003s


## Python

In [ ]:
%%writefile max_pairwise_product_python.py

def max_pairwise_product_python(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    result = max_pairwise_product_python(numbers)
    print(result)

Overwriting max_pairwise_product_python.py


In [ ]:
# verify input14.txt
%%writefile WriteTest_with_python.py

with open("input14.txt", "w") as file:
    n = 11000
    file.write(str(n) + "\n")
    for i in range(1, n + 1):
        file.write(str(i) + " ")

Overwriting WriteTest_with_python.py


In [ ]:
!python WriteTest_with_python.py

In [ ]:
!time python max_pairwise_product_python.py < input14.txt

120989000

real	0m21.244s
user	0m20.930s
sys	0m0.018s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound

Overwriting max_pairwise_product_long.cpp


In [ ]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    print(max_pairwise_product(numbers))

Overwriting max_pairwise_product_python.py


In [ ]:
%%writefile max_pairwise_product_long_python.py
def max_pairwise_product(numbers):
    max_product = 0
    aux_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    print(max_pairwise_product(numbers))


Overwriting max_pairwise_product_long_python.py


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [ ]:
!time python max_pairwise_product_long_python.py < input3.txt

40000000000

real	0m0.047s
user	0m0.035s
sys	0m0.009s


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile sort_solution_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    // 1 instruction
    int n;
    // 1 instruction
    cin >> n;
    // n instructions
    vector<int> numbers(n);

    // 1 + 2n instructions
    for (int i = 0; i < n; ++i) {
        // 1 instruction
        cin >> numbers[i];
    }

    // n*log(n) instructions
    sort(numbers.begin(), numbers.end());
    // 1 instruction
    cout << "Product: " << numbers[n - 1] * numbers[n - 2] << "\n";
    // 1 instruction
    return 0;
}

Writing sort_solution_nlogn.cpp


In [ ]:
%%writefile sort_solution_nlogn.py

n = int(input())  # 1
numbers = list(map(int, input().split()))  # n

numbers.sort()  # n*log(n)
product = numbers[-1] * numbers[-2]  # 1

print(f"Product: {product}")  # 1

Writing sort_solution_nlogn.py


Test with inputs 1 and 2



In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn
!./sort_solution_nlogn < input1.txt
!./sort_solution_nlogn < input2.txt

Product: 90
Product: 90


In [ ]:
!python sort_solution_nlogn.py < input1.txt
!python sort_solution_nlogn.py < input2.txt

Product: 90
Product: 90


Test and time it with input 3

In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn
!time ./sort_solution_nlogn < input3.txt

Product: 1345294336

real	0m0.003s
user	0m0.001s
sys	0m0.001s


In [ ]:
!time python sort_solution_nlogn.py < input3.txt

Product: 40000000000

real	0m0.051s
user	0m0.035s
sys	0m0.013s


20 stress tests

In [ ]:
%%writefile stress_test_x20.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int i = 1; i <= 20; ++i) {
        string fileName = "sinput" + to_string(i) + ".txt";

        ofstream file(fileName);
        if (file.is_open()) {
            int n = rand() % 38 + 3; // Generar un número aleatorio entre 3 y 40
            file << n << "\n";
            for (int j = 0; j < n; ++j) {
                file << rand() % 100 + 1 << " "; // Generar números aleatorios entre 1 y 100
            }
            file.close();
        } else {
            cout << "No se pudo abrir el archivo " << fileName << endl;
        }
    }

    return 0;
}

Overwriting stress_test_x20.cpp


In [ ]:
!g++ stress_test_x20.cpp -o stress_test_x20
!./stress_test_x20

In [ ]:
!./sort_solution_nlogn < sinput1.txt
!./sort_solution_nlogn < sinput2.txt
!./sort_solution_nlogn < sinput3.txt
!./sort_solution_nlogn < sinput4.txt
!./sort_solution_nlogn < sinput5.txt
!./sort_solution_nlogn < sinput6.txt
!./sort_solution_nlogn < sinput7.txt
!./sort_solution_nlogn < sinput8.txt
!./sort_solution_nlogn < sinput9.txt
!./sort_solution_nlogn < sinput10.txt
!./sort_solution_nlogn < sinput11.txt
!./sort_solution_nlogn < sinput12.txt
!./sort_solution_nlogn < sinput13.txt
!./sort_solution_nlogn < sinput14.txt
!./sort_solution_nlogn < sinput15.txt
!./sort_solution_nlogn < sinput16.txt
!./sort_solution_nlogn < sinput17.txt
!./sort_solution_nlogn < sinput18.txt
!./sort_solution_nlogn < sinput19.txt
!./sort_solution_nlogn < sinput20.txt

Product: 9801
Product: 9506
Product: 9800
Product: 9800
Product: 4514
Product: 9215
Product: 5115
Product: 8742
Product: 10000
Product: 9310
Product: 8835
Product: 8613
Product: 2940
Product: 8091
Product: 8910
Product: 9900
Product: 9408
Product: 3608
Product: 9215
Product: 8272


In [ ]:
!python sort_solution_nlogn.py < sinput1.txt
!python sort_solution_nlogn.py < sinput2.txt
!python sort_solution_nlogn.py < sinput3.txt
!python sort_solution_nlogn.py < sinput4.txt
!python sort_solution_nlogn.py < sinput5.txt
!python sort_solution_nlogn.py < sinput6.txt
!python sort_solution_nlogn.py < sinput7.txt
!python sort_solution_nlogn.py < sinput8.txt
!python sort_solution_nlogn.py < sinput9.txt
!python sort_solution_nlogn.py < sinput10.txt
!python sort_solution_nlogn.py < sinput11.txt
!python sort_solution_nlogn.py < sinput12.txt
!python sort_solution_nlogn.py < sinput13.txt
!python sort_solution_nlogn.py < sinput14.txt
!python sort_solution_nlogn.py < sinput15.txt
!python sort_solution_nlogn.py < sinput16.txt
!python sort_solution_nlogn.py < sinput17.txt
!python sort_solution_nlogn.py < sinput18.txt
!python sort_solution_nlogn.py < sinput19.txt
!python sort_solution_nlogn.py < sinput20.txt

Product: 9801
Product: 9506
Product: 9800
Product: 9800
Product: 4514
Product: 9215
Product: 5115
Product: 8742
Product: 10000
Product: 9310
Product: 8835
Product: 8613
Product: 2940
Product: 8091
Product: 8910
Product: 9900
Product: 9408
Product: 3608
Product: 9215
Product: 8272


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile sort_solution_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int main() {
    int instructions;
    // 1 instruction
    int n;
    instructions++;
    // 1 instruction
    cin >> n;
    instructions++;
    // n instructions
    vector<int> numbers(n);
    instructions += n;

    // 1 + 2n instructions
    for (int i = 0; i < n; ++i) {
        // 1 instruction
        instructions++;
        cin >> numbers[i];
        instructions += 2;
    }

    // n*log(n) instructions
    sort(numbers.begin(), numbers.end());
    instructions += n * log2(n);
    // 1 instruction
    cout << "Product: " << numbers[n - 1] * numbers[n - 2] << "\n";
    instructions += 3;
    // 1 instruction
    cout << "Total instructions executed: " << instructions << "\n";
    return 0;
}

Overwriting sort_solution_nlogn.cpp


In [ ]:
!g++ sort_solution_nlogn.cpp -o sort_solution_nlogn

In [ ]:
!./sort_solution_nlogn < input1.txt
!./sort_solution_nlogn < input2.txt
!./sort_solution_nlogn < input3.txt

Product: 90
Total instructions executed: 78
Product: 90
Total instructions executed: 78
Product: 1345294336
Total instructions executed: 44


In [ ]:
%%writefile sort_solution_nlogn.py
import math as m

instructions = 0
n = int(input())  # 1
instructions += 1
numbers = list(map(int, input().split()))  # n
instructions += 4 * n

numbers.sort()  # n*log(n)
instructions += m.log2(n)
product = numbers[-1] * numbers[-2]  # 1
instructions += 2

print(f"Product: {product}")  # 1
instructions += 1

print("Total instructions executed:", int(instructions))

Overwriting sort_solution_nlogn.py


In [ ]:
!python sort_solution_nlogn.py < input1.txt
!python sort_solution_nlogn.py < input2.txt
!python sort_solution_nlogn.py < input3.txt

Product: 90
Total instructions executed: 47
Product: 90
Total instructions executed: 47
Product: 40000000000
Total instructions executed: 30


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!./sort_solution_nlogn < input1.txt

Product: 90
Total instructions executed: 78
CPU times: user 6.44 ms, sys: 1.08 ms, total: 7.52 ms
Wall time: 107 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00644 / 78
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 6.858974358974359e-05
Instructions by second: 14579.43925233645


In [ ]:
%%time
!python sort_solution_nlogn.py < input1.txt

Product: 90
Total instructions executed: 47
CPU times: user 6.33 ms, sys: 1.22 ms, total: 7.55 ms
Wall time: 113 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00633 / 47
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.0001346808510638298
Instructions by second: 7424.960505529226


In [ ]:
%%time
!./sort_solution_nlogn < input2.txt

Product: 90
Total instructions executed: 78
CPU times: user 5.56 ms, sys: 0 ns, total: 5.56 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00556 / 78
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 7.128205128205128e-05
Instructions by second: 14028.776978417267


In [ ]:
%%time
!python sort_solution_nlogn.py < input2.txt

Product: 90
Total instructions executed: 47
CPU times: user 7.26 ms, sys: 0 ns, total: 7.26 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00726 / 47
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00015446808510638299
Instructions by second: 6473.829201101928


In [ ]:
%%time
!./sort_solution_nlogn < input3.txt

Product: 1345294336
Total instructions executed: 44
CPU times: user 5.02 ms, sys: 0 ns, total: 5.02 ms
Wall time: 104 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00502 / 44
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00011409090909090909
Instructions by second: 8764.940239043824


In [ ]:
%%time
!python sort_solution_nlogn.py < input3.txt

Product: 40000000000
Total instructions executed: 30
CPU times: user 6.69 ms, sys: 0 ns, total: 6.69 ms
Wall time: 107 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00669 / 30
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.000223
Instructions by second: 4484.304932735426


### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile sort_efficient_solution.cpp

#include <iostream>
#include <vector>
#include <climits>

using namespace std;

pair<int, int> largest_numbers(vector<int>& numbers) {
    int largest = INT_MIN; // 1
    int secondlargest = INT_MIN; // 1

    for (int i = 0; i < numbers.size(); ++i) { // 1 + 2n
        int act = numbers[i]; // n

        if (act > largest){ // n
            secondlargest = largest; // [0, n]
            largest = act; // [0, n]

        } else if (act > secondlargest) { // n
            secondlargest = act; // [0, n]
        }

    }

    return make_pair(largest, secondlargest);
}

int main() {
    int n; // 1
    cin >> n; // 1

    vector<int> numbers(n); // n

    for (int i = 0; i < n; ++i) { // 1 + 2n
        cin >> numbers[i]; // n
    }

    pair<int, int> result = largest_numbers(numbers); // 1

    cout <<  "Product: " << result.first * result.second << "\n"; // 1
    return 0; // 1
}

Overwriting sort_efficient_solution.cpp


In [ ]:
!g++ sort_efficient_solution.cpp -o sort_efficient_solution

In [ ]:
%%writefile sort_efficient_solution_python.py

def largest_numbers(numbers):
    largest = float('-inf')  # 1
    second_largest = float('-inf')  # 1

    for num in numbers:  # 1 + 2n
        if num > largest:  # n
            second_largest = largest  # [0, n]
            largest = num  # [0, n]
        elif num > second_largest:  # n
            second_largest = num  # [0, n]

    return largest, second_largest

def main():
    n = int(input())  # 1
    numbers = list(map(int, input().split()))  # n

    result = largest_numbers(numbers)  # 1

    print("Product:", result[0] * result[1])  # 1

if __name__ == "__main__":
    main()

Writing sort_efficient_solution_python.py


Test with inputs 1 and 2



In [ ]:
!./sort_efficient_solution < input1.txt
!./sort_efficient_solution < input2.txt

Product: 90
Product: 90


In [ ]:
! python sort_efficient_solution_python.py < input1.txt
! python sort_efficient_solution_python.py < input2.txt

Product: 90
Product: 90


Test and time it with input 3

In [ ]:
!g++ sort_efficient_solution.cpp -o sort_efficient_solution
!time ./sort_efficient_solution < input3.txt

Product: 1345294336

real	0m0.003s
user	0m0.001s
sys	0m0.001s


In [ ]:
!time python sort_efficient_solution_python.py < input3.txt

Product: 40000000000

real	0m0.056s
user	0m0.039s
sys	0m0.011s


20 stress tests

In [ ]:
%%writefile stress_test_x20.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int i = 1; i <= 20; ++i) {
        string fileName = "sinput" + to_string(i) + ".txt";

        ofstream file(fileName);
        if (file.is_open()) {
            int n = rand() % 38 + 3; // Generar un número aleatorio entre 3 y 40
            file << n << "\n";
            for (int j = 0; j < n; ++j) {
                file << rand() % 100 + 1 << " "; // Generar números aleatorios entre 1 y 100
            }
            file.close();
        } else {
            cout << "No se pudo abrir el archivo " << fileName << endl;
        }
    }

    return 0;
}

Overwriting stress_test_x20.cpp


In [ ]:
!g++ stress_test_x20.cpp -o stress_test_x20
!./stress_test_x20

In [ ]:
!./sort_efficient_solution < sinput1.txt
!./sort_efficient_solution < sinput2.txt
!./sort_efficient_solution < sinput3.txt
!./sort_efficient_solution < sinput4.txt
!./sort_efficient_solution < sinput5.txt
!./sort_efficient_solution < sinput6.txt
!./sort_efficient_solution < sinput7.txt
!./sort_efficient_solution < sinput8.txt
!./sort_efficient_solution < sinput9.txt
!./sort_efficient_solution < sinput10.txt
!./sort_efficient_solution < sinput11.txt
!./sort_efficient_solution < sinput12.txt
!./sort_efficient_solution < sinput13.txt
!./sort_efficient_solution < sinput14.txt
!./sort_efficient_solution < sinput15.txt
!./sort_efficient_solution < sinput16.txt
!./sort_efficient_solution < sinput17.txt
!./sort_efficient_solution < sinput18.txt
!./sort_efficient_solution < sinput19.txt
!./sort_efficient_solution < sinput20.txt

Product: 6468
Product: 9216
Product: 8190
Product: 9405
Product: 9700
Product: 3325
Product: 8439
Product: 9024
Product: 8645
Product: 8100
Product: 8178
Product: 8544
Product: 5254
Product: 7990
Product: 9801
Product: 8742
Product: 9900
Product: 8554
Product: 6992
Product: 10000


In [ ]:
!python sort_efficient_solution_python.py < sinput1.txt
!python sort_efficient_solution_python.py < sinput2.txt
!python sort_efficient_solution_python.py < sinput3.txt
!python sort_efficient_solution_python.py < sinput4.txt
!python sort_efficient_solution_python.py < sinput5.txt
!python sort_efficient_solution_python.py < sinput6.txt
!python sort_efficient_solution_python.py < sinput7.txt
!python sort_efficient_solution_python.py < sinput8.txt
!python sort_efficient_solution_python.py < sinput9.txt
!python sort_efficient_solution_python.py < sinput10.txt
!python sort_efficient_solution_python.py < sinput11.txt
!python sort_efficient_solution_python.py < sinput12.txt
!python sort_efficient_solution_python.py < sinput13.txt
!python sort_efficient_solution_python.py < sinput14.txt
!python sort_efficient_solution_python.py < sinput15.txt
!python sort_efficient_solution_python.py < sinput16.txt
!python sort_efficient_solution_python.py < sinput17.txt
!python sort_efficient_solution_python.py < sinput18.txt
!python sort_efficient_solution_python.py < sinput19.txt
!python sort_efficient_solution_python.py < sinput20.txt

Product: 6468
Product: 9216
Product: 8190
Product: 9405
Product: 9700
Product: 3325
Product: 8439
Product: 9024
Product: 8645
Product: 8100
Product: 8178
Product: 8544
Product: 5254
Product: 7990
Product: 9801
Product: 8742
Product: 9900
Product: 8554
Product: 6992
Product: 10000


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile sort_efficient_solution.cpp

#include <iostream>
#include <vector>
#include <climits>

using namespace std;
int instructions;

pair<int, int> largest_numbers(vector<int>& numbers) {
    int largest = INT_MIN; // 1
    instructions++;
    int secondlargest = INT_MIN; // 1
    instructions++;

    for (int i = 0; i < numbers.size(); ++i) { // 1 + 2n
        int act = numbers[i]; // n
        instructions += 2;

        if (act > largest){ // n
            secondlargest = largest; // [0, n]
            largest = act; // [0, n]
            instructions += 2;

        } else if (act > secondlargest) { // n
            secondlargest = act; // [0, n]
            instructions ++;
        }

    }

    return make_pair(largest, secondlargest);
}

int main() {
    int n; // 1
    instructions++;
    cin >> n; // 1
    instructions++;

    vector<int> numbers(n); // n
    instructions += n;

    for (int i = 0; i < n; ++i) { // 1 + 2n
        cin >> numbers[i]; // n
        instructions++;
    }

    pair<int, int> result = largest_numbers(numbers); // 1
    instructions++;

    cout <<  "Product: " << result.first * result.second << "\n"; // 1
    cout << "Total instructions executed: " << instructions << "\n";
    return 0; // 1
}

Overwriting sort_efficient_solution.cpp


In [ ]:
!g++ sort_efficient_solution.cpp -o sort_efficient_solution

In [ ]:
!./sort_efficient_solution < input1.txt
!./sort_efficient_solution < input2.txt
!./sort_efficient_solution < input3.txt

Product: 90
Total instructions executed: 65
Product: 90
Total instructions executed: 48
Product: 1345294336
Total instructions executed: 37


In [ ]:
%%writefile sort_efficient_solution_python.py

def largest_numbers(numbers):
    instructions = 0
    largest = float('-inf')  # 1
    instructions += 1
    second_largest = float('-inf')  # 1
    instructions += 1

    for num in numbers:  # 1 + 2n
        if num > largest:  # n
            second_largest = largest  # [0, n]
            largest = num  # [0, n]
            instructions += 2
        elif num > second_largest:  # n
            second_largest = num  # [0, n]
            instructions += 1

    return largest, second_largest, instructions

def main():
    instructions = 0

    n = int(input())  # 1
    instructions += 2
    numbers = list(map(int, input().split()))  # n
    instructions += 4*n

    result = largest_numbers(numbers)  # 1
    instructions += result[2] + 1

    print("Product:", result[0] * result[1])  # 1
    print("Total instructions executed:", instructions)

if __name__ == "__main__":
    main()

Overwriting sort_efficient_solution_python.py


In [ ]:
!python sort_efficient_solution_python.py < input1.txt
!python sort_efficient_solution_python.py < input2.txt
!python sort_efficient_solution_python.py < input3.txt

Product: 90
Total instructions executed: 65
Product: 90
Total instructions executed: 48
Product: 40000000000
Total instructions executed: 37


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!./sort_efficient_solution < input1.txt

Product: 90
Total instructions executed: 65
CPU times: user 3.96 ms, sys: 1.2 ms, total: 5.17 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00396 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 6.092307692307692e-05
Instructions by second: 16414.141414141413


In [ ]:
%%time
!python sort_efficient_solution_python.py < input1.txt

Product: 90
Total instructions executed: 65
CPU times: user 4.79 ms, sys: 917 µs, total: 5.71 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00479 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 7.36923076923077e-05
Instructions by second: 13569.937369519832


In [ ]:
%%time
!./sort_efficient_solution < input2.txt

Product: 90
Total instructions executed: 48
CPU times: user 5.13 ms, sys: 1.18 ms, total: 6.31 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00513 / 48
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.000106875
Instructions by second: 9356.72514619883


In [ ]:
%%time
!python sort_efficient_solution_python.py < input2.txt

Product: 90
Total instructions executed: 48
CPU times: user 5.49 ms, sys: 0 ns, total: 5.49 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00549 / 48
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.000114375
Instructions by second: 8743.169398907105


In [ ]:
%%time
!./sort_efficient_solution < input3.txt

Product: 1345294336
Total instructions executed: 37
CPU times: user 5.7 ms, sys: 831 µs, total: 6.53 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.0057 / 37
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00015405405405405405
Instructions by second: 6491.228070175439


In [ ]:
%%time
!python sort_efficient_solution_python.py < input3.txt

Product: 40000000000
Total instructions executed: 37
CPU times: user 5.84 ms, sys: 30 µs, total: 5.87 ms
Wall time: 106 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00584 / 37
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 0.00015783783783783782
Instructions by second: 6335.616438356165


### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile super_efficient_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    long long totalInstr = 0LL;

    int n;
    cin >> n;
    totalInstr += 3;

    long long largest = 0LL;
    long long secondLargest = 0LL;

    for (int i = 0; i < n; i++) {
        int num;
        cin >> num;
        totalInstr += 2;

        if (num >= largest) {
            secondLargest = largest;
            largest = num;
            totalInstr += 4;
        } else if (num > secondLargest) {
            secondLargest = num;
            totalInstr += 2;
        }
    }

    long long result = largest * secondLargest;
    totalInstr++;

    cout << "Total Instructions: " << totalInstr << endl;
    cout << "Result: " << result << endl;

    return 0;
}

Overwriting super_efficient_solution.cpp


In [ ]:
!g++ ./super_efficient_solution.cpp -o ./super_efficient_solution

Test with inputs 1 and 2




In [ ]:
!./super_efficient_solution < input1.txt
!./super_efficient_solution < input2.txt

Total Instructions: 28
Result: 64
Total Instructions: 4
Result: 0


Test and time it with input 3

In [ ]:
%%time
!./super_efficient_solution < input3.txt

Total Instructions: 34
Result: 40000000000
CPU times: user 5.69 ms, sys: 0 ns, total: 5.69 ms
Wall time: 105 ms


20 stress tests

In [ ]:
%%writefile super_stress_test_x20.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int i = 1; i <= 20; ++i) {
        string fileName = "sinput" + to_string(i) + ".txt";

        ofstream file(fileName);
        if (file.is_open()) {
            int n = rand() % 38 + 3; // Generar un número aleatorio entre 3 y 40
            file << n << " " << n << "\n"; // Escribir dos números aleatorios en la primera línea
            for (int j = 0; j < n; ++j) {
                file << rand() % 100 + 1 << " "; // Generar números aleatorios entre 1 y 100
            }
            file.close();
        } else {
            cout << "No se pudo abrir el archivo " << fileName << endl;
        }
    }

    return 0;
}


Overwriting super_stress_test_x20.cpp


In [ ]:
!g++ super_stress_test_x20.cpp -o super_stress_test_x20
!./super_stress_test_x20

In [ ]:
!./super_efficient_solution < sinput1.txt
!./super_efficient_solution < sinput2.txt
!./super_efficient_solution < sinput3.txt
!./super_efficient_solution < sinput4.txt
!./super_efficient_solution < sinput5.txt
!./super_efficient_solution < sinput6.txt
!./super_efficient_solution < sinput7.txt
!./super_efficient_solution < sinput8.txt
!./super_efficient_solution < sinput9.txt
!./super_efficient_solution < sinput10.txt
!./super_efficient_solution < sinput11.txt
!./super_efficient_solution < sinput12.txt
!./super_efficient_solution < sinput13.txt
!./super_efficient_solution < sinput14.txt
!./super_efficient_solution < sinput15.txt
!./super_efficient_solution < sinput16.txt
!./super_efficient_solution < sinput17.txt
!./super_efficient_solution < sinput18.txt
!./super_efficient_solution < sinput19.txt
!./super_efficient_solution < sinput20.txt

Total Instructions: 58
Result: 9108
Total Instructions: 90
Result: 9600
Total Instructions: 44
Result: 4345
Total Instructions: 30
Result: 3520
Total Instructions: 92
Result: 9405
Total Instructions: 70
Result: 6536
Total Instructions: 76
Result: 9312
Total Instructions: 48
Result: 9207
Total Instructions: 102
Result: 9800
Total Instructions: 66
Result: 9009
Total Instructions: 86
Result: 8372
Total Instructions: 112
Result: 8836
Total Instructions: 92
Result: 8008
Total Instructions: 48
Result: 5313
Total Instructions: 104
Result: 9800
Total Instructions: 54
Result: 6474
Total Instructions: 94
Result: 9306
Total Instructions: 86
Result: 9108
Total Instructions: 52
Result: 8648
Total Instructions: 46
Result: 8084


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
!./super_efficient_solution < input1.txt
!./super_efficient_solution < input2.txt
!./super_efficient_solution < input3.txt

Total Instructions: 28
Result: 64
Total Instructions: 4
Result: 0
Total Instructions: 34
Result: 40000000000


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!./super_efficient_solution < input1.txt

Total Instructions: 28
Result: 64
CPU times: user 3.22 ms, sys: 1.93 ms, total: 5.14 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00322 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 4.953846153846154e-05
Instructions by second: 20186.335403726705


In [ ]:
%%time
!./super_efficient_solution < input2.txt

Total Instructions: 4
Result: 0
CPU times: user 5.85 ms, sys: 0 ns, total: 5.85 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00585 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 9e-05
Instructions by second: 11111.111111111111


In [ ]:
%%time
!./super_efficient_solution < input3.txt

Total Instructions: 34
Result: 40000000000
CPU times: user 4.8 ms, sys: 1.02 ms, total: 5.82 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00480 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 7.384615384615384e-05
Instructions by second: 13541.666666666666


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile super_efficient_solution_python.py
totalInstr = 0

n, m = map(int, input().split())  # Lee dos números de la primera línea
totalInstr += 3

largest = 0
secondLargest = 0

for i in range(n):
    num = i
    totalInstr += 2

    if num >= largest:
        secondLargest = largest
        largest = num
        totalInstr += 4
    elif num > secondLargest:
        secondLargest = num
        totalInstr += 2

result = largest * secondLargest
totalInstr += 1

print("Total Instructions:", totalInstr)
print("Result:", result)

Overwriting super_efficient_solution_python.py


Test with inputs 1 and 2




In [ ]:
!python super_efficient_solution_python.py < input1.txt
!python super_efficient_solution_python.py < input2.txt

Total Instructions: 28
Result: 6
Total Instructions: 4
Result: 0


Test and time it with input 3

In [ ]:
%%time
!./super_efficient_solution_python.py < input3.txt

/bin/bash: line 1: ./super_efficient_solution_python.py: Permission denied
CPU times: user 6.38 ms, sys: 1.21 ms, total: 7.59 ms
Wall time: 112 ms


20 stress tests

In [ ]:
%%writefile super_stress_test_x20_python.py
import random

for i in range(1, 21):
    fileName = f"sinput{i}.txt"

    with open(fileName, "w") as file:
        n = random.randint(3, 40)  # Generar un número aleatorio entre 3 y 40
        file.write(f"{n} {n}\n")  # Escribir dos números aleatorios en la primera línea
        for j in range(n):
            number = random.randint(1, 100)  # Generar números aleatorios entre 1 y 100
            file.write(f"{number} ")

print("Archivos de entrada generados.")

Overwriting super_stress_test_x20_python.py


In [ ]:
!python super_stress_test_x20_python.py

Archivos de entrada generados.


In [ ]:
!python super_efficient_solution_python.py < sinput1.txt
!python super_efficient_solution_python.py < sinput2.txt
!python super_efficient_solution_python.py < sinput3.txt
!python super_efficient_solution_python.py < sinput4.txt
!python super_efficient_solution_python.py < sinput5.txt
!python super_efficient_solution_python.py < sinput6.txt
!python super_efficient_solution_python.py < sinput7.txt
!python super_efficient_solution_python.py < sinput8.txt
!python super_efficient_solution_python.py < sinput9.txt
!python super_efficient_solution_python.py < sinput10.txt
!python super_efficient_solution_python.py < sinput11.txt
!python super_efficient_solution_python.py < sinput12.txt
!python super_efficient_solution_python.py < sinput13.txt
!python super_efficient_solution_python.py < sinput14.txt
!python super_efficient_solution_python.py < sinput15.txt
!python super_efficient_solution_python.py < sinput16.txt
!python super_efficient_solution_python.py < sinput17.txt
!python super_efficient_solution_python.py < sinput18.txt
!python super_efficient_solution_python.py < sinput19.txt
!python super_efficient_solution_python.py < sinput20.txt

Total Instructions: 130
Result: 380
Total Instructions: 154
Result: 552
Total Instructions: 100
Result: 210
Total Instructions: 58
Result: 56
Total Instructions: 58
Result: 56
Total Instructions: 118
Result: 306
Total Instructions: 136
Result: 420
Total Instructions: 142
Result: 462
^C
Total Instructions: 22
Result: 2
Total Instructions: 106
Result: 240
Total Instructions: 166
Result: 650
Total Instructions: 166
Result: 650
Total Instructions: 100
Result: 210
Total Instructions: 64
Result: 72
Total Instructions: 190
Result: 870
Total Instructions: 28
Result: 6
Total Instructions: 106
Result: 240
Total Instructions: 220
Result: 1190
Total Instructions: 160
Result: 600


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
!python super_efficient_solution_python.py < input1.txt
!python super_efficient_solution_python.py < input2.txt
!python super_efficient_solution_python.py < input3.txt

Total Instructions: 28
Result: 6
Total Instructions: 4
Result: 0
Traceback (most recent call last):
  File "/content/super_efficient_solution_python.py", line 3, in <module>
    n, m = map(int, input().split())  # Lee dos números de la primera línea
ValueError: not enough values to unpack (expected 2, got 1)


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!python super_efficient_solution_python.py < input1.txt

Total Instructions: 28
Result: 6
CPU times: user 5.91 ms, sys: 1.02 ms, total: 6.93 ms
Wall time: 107 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00519 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 7.984615384615386e-05
Instructions by second: 12524.084778420038


In [ ]:
%%time
!python super_efficient_solution_python.py < input2.txt

Total Instructions: 4
Result: 0
CPU times: user 5.57 ms, sys: 976 µs, total: 6.54 ms
Wall time: 105 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00600 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 9.230769230769232e-05
Instructions by second: 10833.333333333332


In [ ]:
%%time
!python super_efficient_solution_python.py < input3.txt

Traceback (most recent call last):
  File "/content/super_efficient_solution_python.py", line 3, in <module>
    n, m = map(int, input().split())  # Lee dos números de la primera línea
ValueError: not enough values to unpack (expected 2, got 1)
CPU times: user 7.79 ms, sys: 1.11 ms, total: 8.91 ms
Wall time: 109 ms


In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00604 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

Time by instruction: 9.292307692307692e-05
Instructions by second: 10761.58940397351
